move route expiration to kernel

dqminh opened the pull request:
netlink API now supports setting route expiration with RTA_EXPIRES
(https://github.com/torvalds/linux/commit/32bc201e1974976b7d3fea9a9b17bb7392ca6394),
so here we remove the current route expiration done in userspace and
depend on kernel to expire routes for us.

Fix #8048

keszybz commented:
That's a beatiful patch ;) But I'm not sure we can merge it as is, because we generally support a much lower kernel version. README currently says \>= 13.3, and the kernel support for this appeared in 4.5. I don't think we should bump this to 4.5 yet (it's quite painful for some users of systemd, especially on embedded an other special environments, when we update the required kernel version). Debian stable has 4.5, oldstable has 3.5 and 4.3 in backports.

So the normal approach would be to use the new functionality (`RTA_EXPIRES` in this case) and fall back to the old code. But that's not so easy, because we set the attribute, and if it's not supported, we'll most likely get an error from the kernel when calling `route_add`, which is quite a bit later... Then we'd need to retry the call without that set, which is doable but very awkward and verbose. I wonder if we should just depend on the kernel version returned by uname(2). Normally we prefer to do functional checks, but in this case it's not so important, we'd just spuriously fall back to the old code in the event that somebody backported this. We have `str_verscmp` already (used for something similar in `src/shared/condition.c`), so this shouldn't take more than a few lines.

@poettering comments on the general direction here?

poettering commented:
yeah, we need some fallback code there to retain compat with older kernels... Not sure how we can do that nicely though. Maybe add some function that tries to add some low-prio route on the loopback interface with the expires attribute set. If that works we cache the result and know that auto-expiry routes work. Either way we'd immediately drop the route again...

it's frickin ugly, but I much prefer if we check for features rather than kernel versions...

yuwata commented on src/network/networkd-route.c:
```diff
@@ -617,6 +579,12 @@ int route_configure(
 
+        if (route->lifetime != SEC_INFINITY) {
+                r = sd_netlink_message_append_u32(req, RTA_EXPIRES, route->lifetime);
+                if (r < 0)
```
When this fails, then we should go to fallback logic, that is using `sd_event_add_time`. What do you think? This way, we can safely support old kernels.

keszybz replied:
FTR, since github is attaching an old comment here: this won't work, because we don't communicate with the kernel yet.

keszybz commented:
> When this fails

This is not going to fail. `route_add` is going to fail, much later and it's also not obvious that this is the reason. E.g. it might fail because some route is illegal, and how do we distinguish that?

> Maybe add some function that tries to add some low-prio route on the loopback 

Meh, that seems so complicated, and really, we know that any kernel >= 4.5 has this, so false positives are not a problem, and if we hit a false negative, we just use a non-optimal approach. So yeah, in this case a simple version check seems to be the best solution.

dqminh commented:
@keszybz @poettering i added a conditional for kernel >= 4.5. Agree that its quite ugly, but checking for RTA_EXPIRES seems too complicated in this case.

keszybz commented on src/network/networkd-manager.c:
```diff
@@ -1444,6 +1446,12 @@ int manager_new(Manager **ret, sd_event *event) {
         (void) routing_policy_load_rules(m->state_file, &m->rules_saved);
 
+        // Router lifetime can be set with netlink interface since kernel >= 4.5
+        // so for the supported kernel we dont need to expire routes in userspace
```
Please use `/* */` comments.

keszybz commented on src/network/networkd-manager.c:
```diff
@@ -1444,6 +1446,12 @@ int manager_new(Manager **ret, sd_event *event) {
 
+        // Router lifetime can be set with netlink interface since kernel >= 4.5
+        // so for the supported kernel we dont need to expire routes in userspace
+        route_expiration_condition = condition_new(CONDITION_KERNEL_VERSION, ">= 4.5", false, false);
```
So, either the result of `condition_new` should be used for oom, and the result of `condition_test` should be check for error, or maybe, which I think is a nicer option, just allocate this statically and assert:
```C
{
         static const Condition c = { .type = CONDITION_KERNEL_VERSION, .parameter = ">= 4.5");

         r = condition_test(route_expiration_condition);
         assert(r >= 0);
         m->route_expiration_kernel = !!r;
}
```

poettering commented on src/network/networkd-manager.c:
```diff
@@ -1444,6 +1445,19 @@ int manager_new(Manager **ret, sd_event *event) {
+         * so for the supported kernel we dont need to expire routes in userspace */
+        {
+          Condition c = { .type = CONDITION_KERNEL_VERSION };
+          c.parameter = strdup(">= 4.5");
```
why strdup this? why not just do: 

```c
Condition c = {
        .type = CONDITION_KERNEL_VERSION,
        .parameter = (char*) ">= 4.5",
};
```

keszybz replied:
... which is what I wrote in my original comment ;)

poettering commented on src/network/networkd-manager.c:
```diff
@@ -1444,6 +1445,19 @@ int manager_new(Manager **ret, sd_event *event) {
+          if (!c.parameter)
+                  return log_oom();
+          r = condition_test(&c);
+          assert(r >= 0);
```
please propagate the error properly instead of asserting on it. This is unlikely to trigger in the current implementation, but maybe the condition stuff is changed later on

poettering commented on src/network/networkd-manager.c:
```diff
@@ -1444,6 +1445,19 @@ int manager_new(Manager **ret, sd_event *event) {
+          assert(r >= 0);
+          m->route_expiration_kernel = !!r;
+          free(c.parameter);
+        }
```
also, please indent by multiples of 8ch.

poettering commented on src/basic/time-util.h:
```diff
@@ -48,6 +48,7 @@ typedef struct triple_timestamp {
         usec_t boottime;
 } triple_timestamp;
 
+#define SEC_INFINITY  ((uint32_t) -1)
```
i don't think we should this macro. The type is too unspecific I fear. People store seconds in any kind of types, not just uint32_t. In fact the ndisc lifetime is actually a uint16_t, hence there's already confusion about the type within this very patch...

Moreover I am pretty sure we should stick to usec_t times wherever appropriate, in particular in generic code, hence the Route's lifetime field should be changed to usec_t even if it is initialized from the ndisc lifetime value (that is if we need the field at all. But i have the suspicion we should drop it altogether since it is redundant)

poettering commented on src/network/networkd-route.c:
```diff
@@ -617,6 +618,12 @@ int route_configure(
                 return log_error_errno(r, "Could not append RTA_PREF attribute: %m");
 
+        if (route->lifetime != SEC_INFINITY && link->manager->route_expiration_kernel) {
+                r = sd_netlink_message_append_u32(req, RTA_EXPIRES, route->lifetime);
```
hmm, i am pretty sure the `lifetime` field of Route should go, and instead we should calculate the expiry here like this:

`DIV_ROUND_UP(usec_sub(route->expire_time, now(clock_boottime_or_monotonic()), USEC_PER_SEC);`

I.e. calculate the remaining time in a saturated way at the moment where we set this field. That way we don't need the liftime field at all

poettering commented on src/network/networkd-manager.c:
```diff
@@ -1444,6 +1445,16 @@ int manager_new(Manager **ret, sd_event *event) {
+                        return r;
+                m->route_expiration_kernel = !!r;
+        }
+
```
instead of initializing this here during initialization time, I think we should determine this on demand only, i.e. introduce a new call like this:

```c
int kernel_route_expiration_supported(void) {
        static int cached = -1;
        int r;

        if (cached < 0) {
                Condition c = { 
                        .type = CONDITION_KERNEL_VERSION, 
                        .parameter = (char *) ">= 4.5" 
                };
                r = condition_test(&c);
                if (r < 0)
                        return r;

                cached = r;
        }
        return cached;
}
```

This  call would cache the result internally and only run the check on the first invocation.

dqminh commented:
@poettering @keszybz Thanks for the reviews ! Updated to latest rev which revert the renaming of lifetime field and just added the relevant bits from latest reviews instead.

poettering commented:
lgtm
